openstudiocore/src/utilities/filetypes/EpwFile.cpp

```diff
@@ -2202,204 +2866,1371 @@ namespace openstudio{
+    std::vector<double> values;
+    for (unsigned int i = 0; i<m_data.size(); i++) {
+      Date date = m_data[i].date();
+      Time time = m_data[i].time();
+      boost::optional<double> value = (m_data[i].*compute)();
+      if (value) {
+        dates.push_back(DateTime(date, time));
+        values.push_back(value.get());
+      }
+    }
+    if (values.size()) {
+      DateTime start = dates[1] - Time(0, 0, 0, 3600.0 / m_recordsPerHour);
+      dates[0] = start; // Overwrite the placeholder
+      return boost::optional<TimeSeries>(TimeSeries(dates, openstudio::createVector(values), units));
+    }
+    return boost::none;
+  }
+
+  bool EpwFile::translateToWth(openstudio::path path, std::string description)
+  {
+    if(m_data.size()==0) {
+      if (!openstudio::filesystem::exists(m_path) || !openstudio::filesystem::is_regular_file(m_path)){
+        LOG_AND_THROW("Path '" << m_path << "' is not an EPW file");
+      }
+
+      // set checksum
+      m_checksum = openstudio::checksum(m_path);
+
+      // open file
+      std::ifstream ifs(openstudio::toString(m_path));
+
+      if (!parse(ifs, true)) {
+        ifs.close();
+        LOG(Error,"EpwFile '" << toString(m_path) << "' cannot be processed");
+        return false;
+      }
+      ifs.close();
+    }
+
+    if(description.empty()) {
+      description = "Translated from " + openstudio::toString(this->path());
+    }
+
+    if(!data().size()) {
+      LOG(Error, "EPW file contains no data to translate");
+      return false;
+    }
+
+    openstudio::filesystem::ofstream fp(path, std::ios_base::binary);
+    if(!fp.is_open()) {
+      LOG(Error, "Failed to open file '" + openstudio::toString(path) + "'");
+      return false;
+    }
+
+    fp << "WeatherFile ContamW 2.0\n";
+    fp << description << "\n";
+    fp << toString(QString("%1/%2\t!start date\n").arg(openstudio::month(startDate().monthOfYear())).arg(startDate().dayOfMonth()));
+    fp << toString(QString("%1/%2\t!end date\n").arg(openstudio::month(endDate().monthOfYear())).arg(endDate().dayOfMonth()));
+    fp << "!Date\tDofW\tDtype\tDST\tTgrnd [K]\n";
+    openstudio::Time delta(1,0);
+    int dayofweek = startDayOfWeek().value()+1;
+    for(openstudio::Date current=startDate();current<=endDate();current += delta) {
+      fp << toString(QString("%1/%2\t%3\t%3\t0\t283.15\n")
+          .arg(openstudio::month(current.monthOfYear()))
+          .arg(current.dayOfMonth())
+          .arg(dayofweek));
+      dayofweek++;
+      if(dayofweek > 7) {
+        dayofweek=1;
+      }
+    }
+
+    // Cheat to get data at the start time - this will need to change
+    openstudio::EpwDataPoint lastPt = data()[data().size()-1];
+    std::vector<std::string> epwstrings = lastPt.toEpwStrings();
+    openstudio::DateTime dateTime = data()[0].dateTime();
+    openstudio::Time dt = timeStep();
+    dateTime -= dt;
+    epwstrings[0] = std::to_string(dateTime.date().year());
```

Comment:
Need to check size before writing to index

Reply:
Not me.